UTRannotator.pm

```diff
@@ -0,0 +1,1306 @@
+    );
+
+
+    my @output_five_prime_flag=();
+    my @output_five_prime_annotation=();
+
+    foreach my $flag (keys %five_prime_flag){
+        if($five_prime_flag{$flag}){
+            push @output_five_prime_flag,$five_prime_flag{$flag};
+            push @output_five_prime_annotation,$five_prime_annotation{$flag};
+        }
+    };
+
+	@output_five_prime_flag=("-") if(!@output_five_prime_flag);
+	@output_five_prime_annotation=("-") if(!@output_five_prime_annotation);
+
+    my %utr_effect = (
+        "five_prime_UTR_variant_consequence" => (join "&", @output_five_prime_flag),
+        "five_prime_UTR_variant_annotation" => (join "&", @output_five_prime_annotation),
+    );
+
+  	my $existing_uORF_num = $self->count_number_ATG($five_prime_seq);
+	my $output ={%utr_effect, %$existing_uORF_num};
+	return $output? $output: {};
+}
+
+##################
+# Main functions #
+##################
+
+
+sub uAUG_gained {
+	# Description: annotate if a five_prime_UTR_variant creates ATG
+
+	my ($self, $variant_info,$UTR_info, %kozak_strength) = @_;
+
+	my $pos = $variant_info->{pos};
+	my $ref = $variant_info->{ref};
+	my $alt = $variant_info->{alt};
+
+	my @sequence = split //, $UTR_info->{seq};
+	my $strand = $UTR_info->{strand};
+
+	#return annotators
+	my $uAUG_gained_DistanceToCDS = "";  # the distance between the gained uAUG to CDS
+	my $uAUG_gained_KozakContext = "";  # the Kozak context sequence of the gained uAUG
+	my $uAUG_gained_KozakStrength = ""; # the Kozak strength of the gained uAUG
+	my $uAUG_gained_type = ""; # the type of uORF created - any of the following: uORF, inframe_oORF,OutOfFrame_oORF
+	my $uAUG_gained_DistanceFromCap = ""; # the distance between the gained uAUG to the start of the five prime UTR
+    my $uAUG_gained_DistanceToStop = ""; #the distance between the gained uAUG to stop codon (could be in CDS)
+
+	#indicate whether the variant creates a ATG
+	my $flag = 0;
+	my $current_kozak = "";
+	my $current_kozak_strength ="";
+
+	#the relative position of input variant in the UTR sequence
+
+    my %result = (); #result is a hash table with two elements: $flag and $output_effects
+    my $output_flag = "";
+    my $output_effects = "";
+
+    my $ref_coding = $self->get_ref_coding($ref);
+    my $alt_coding = $alt eq "-"? "" : $alt;
+    reverse_comp(\$alt_coding) if($strand < 0);
+
+	my ($mut_pos, $end_pos) = $self->get_allele_exon_pos($strand, $pos, $ref_coding, $UTR_info);
+    return {} unless(defined($mut_pos)&defined($end_pos));
+
+	my $mut_utr_seq = $self->mut_utr_sequence(\@sequence,$mut_pos,$ref_coding,$alt_coding,$strand);
+  	my @mut_utr_seq = split //,$mut_utr_seq;
+    my $mut_utr_length = @mut_utr_seq;
+
+    #get the nt sequence that might have the pos_A for a new ATG codon
+    my @mut_seq = @mut_utr_seq[$mut_pos-2..$mut_pos+length($alt_coding)+1];
+    my @mut_atg_pos = @{$self->get_ATG_pos(\@mut_seq)};
+
+    # check whether there is a ATG codon is to check the length of the array
```

Comment:
I don't understand this sentence, does it mean: check whether there is a ATG codon, this codon will be used to check the length of the array?

Reply:
I  think what it means is: "Check if ATG codon exists. In case, it does exists it will give length relative to A position."